portfolio/src/main/java/com/google/sps/servlets/UploadImageServlet.java

```diff
@@ -0,0 +1,104 @@
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.QueryResultList;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/upload-image")
+public class UploadImageServlet extends HttpServlet {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        String imageUrl = getUploadedFileUrl(request, "image");
+        if(imageUrl == null) return;
+        UserService userService = UserServiceFactory.getUserService();
+        String id = userService.getCurrentUser().getUserId();
+        Entity userInfoEntity = Utils.getEntity(id);
+        userInfoEntity.setProperty("profilePictureUrl", imageUrl);
+        datastore.put(userInfoEntity);
+        response.sendRedirect("/profile.html");
+    }
+
+    /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file. */
+    private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+        BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+        Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+        List<BlobKey> blobKeys = blobs.get(formInputElementName);
+
+        // User submitted form without selecting a file, so we can't get a URL. (dev server)
+        if (blobKeys == null || blobKeys.isEmpty()) {
+            return null;
+        }
+
+        // Our form only contains a single file input, so get the first index.
+        BlobKey blobKey = blobKeys.get(0);
+
+        // User submitted form without selecting a file, so we can't get a URL. (live server)
+        BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+        if (blobInfo.getSize() == 0) {
+            blobstoreService.delete(blobKey);
+            return null;
+        }
+
+        // We could check the validity of the file here, e.g. to make sure it's an image file
+        // https://stackoverflow.com/q/10779564/873165
+
+        // Use ImagesService to get a URL that points to the uploaded file.
+        ImagesService imagesService = ImagesServiceFactory.getImagesService();
+        ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+
+        // To support running in Google Cloud Shell with AppEngine's dev server, we must use the relative
+        // path to the image, rather than the path returned by imagesService which contains a host.
+        try {
+            URL url = new URL(imagesService.getServingUrl(options));
```

Comment:
The API "ImageService#getServingUrl(ServingUrlOptions)" can throws the other two exceptions java.lang.IllegalArgumentException and ImagesServiceFailureException. Could you please point out to me which line could cause the MalformedURLException?

https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/images/ImagesService.html#getServingUrl-com.google.appengine.api.images.ServingUrlOptions-

Also, we would like to put the try-catch block as narrow as possible. So people knows which part of codes might throw exceptions. So I will suggest to replace the code like below

```java
String servingUrl = null;
try {
  servingUrl = imagesService.getServingUrl(options);
} catch (IllegalArgumentException | ImagesServiceFailureException e) {
  // Do nothing here. But we might add some logs for debugging.
}

// servingUrl will remains null if any of the above exceptions occur.
if (servingUrl == null) {
    return null;
}

URL url = new URL(servingUrl);
return url.getPath();
```

Since this method might return null String, you probably need to handle it outside of this class.

Reply:
I have no idea how this part of the code works. It was given in one of the examples and I copied it from there. Also, it won't compile if I don't catch MalformedURLException, so this is actually necessary.